### ISSUE TITLE
Make interesting conversations about code easier to find?

### ISSUE THREAD

**Issue by stoft** [exercism/v2-feedback#124]
Hi,

Sorry if this is a duplicate or similar, I searched a bit, couldn't find anything and my time is too limited to research more before posting.

I'm an old exercism user who loved it for a few months three years ago and since then have only used it sparingly, mostly due to other priorities in life but also due to spending time on a coding project of mine instead of coding exercises. I do still recommend it far and wide whenever I get the chance.

Today a new exercism user who had seen our conversations from three years ago and found them really valuable took the time to hunt us down on Slack and personally communicate his appreciation:

> I just wanted to thank each of you for the wonderful conversations you've had around solving a number of the Elixir exercises on Exercism. My solution (...) for say the `sublist` exercise is inspired by each of your solutions in different ways. Thank you once again as your conversations make learning Elixir even more exciting for me through Exercism as I can look forward to more of these learnings (through conversations around each iteration) each time I pick up a new exercise :slightly_smiling_face:. Please do keep these conversations going whenever you solve more exercises in any language!

This made me realise that one of the best things about exercism to me was, and still is the fantastic conversations people have about code. My personal experience is also that these conversations do not always occur on Exercism and my worry is that with thousands of solutions to the same problem the interesting conversations become hard to find. Moving forward into v2 it would be nice if these conversations could be amplified somehow so that they become easier to find, e.g. for people who join the community years after the conversations happened. Perhaps this diverges from the goals of v2 but I couldn't stop myself from at least providing the feedback.

Regardless, a big thank you for making exercism!

**Comment by boddhisattva** [comment]
Thanks for opening up this discussion @stoft 👍 . I actually have a related question, even in the current platform, what is the best way of finding those solutions for a given language that have more conversations around it and do we have something already as a feature(or something that is already being worked upon ?) that addresses this in Nextercism ?

I most likely end up looking at those solutions that have more conversations around it as a first preference because those conversations are talking about things like trade offs, thought process behind why they use `x` wrt a given language and so on. I've found identifying what things I've not considered when approaching a given problem and learning from that experience as I go about solving new problems is something I really care about because although I can learn a good amount of stuff from a given solution, it's the journey to get to that solution is what makes me improve my thinking and approach towards problem solving via a given language and as a programmer in general.

**Comment by boddhisattva** [comment]
This might sound like a shameless plug, but in [this post](https://codeburst.io/the-journey-of-solving-an-elixir-exercise-on-exercism-and-how-that-led-me-to-look-beyond-the-98fb9d7ae298) I've made an attempt to capture how much I've benefitted from thoughtful conversations about code  and I'm sharing this here because it just echoes how useful finding interesting conversations about code(in the context of the above github issue) can be! This experience truly was one instance of Exercism being at it's best for me! Thanks Katrina and everyone else who've helped out with Exercism.

**Comment by kytrinyx** [comment]
> my worry is that with thousands of solutions to the same problem the interesting conversations become hard to find. Moving forward into v2 it would be nice if these conversations could be amplified somehow so that they become easier to find

I'm so sorry I missed this when you posted it @stoft!

There are so many interesting and hard problems and trade-offs tucked into this. Surfacing great conversations has long been one of the things I hope to tackle once we get the basic usability of the site worked out.

**Comment by stoft** [comment]
@kytrinyx no worries, you've seen it which is the important part. :) Happy Christmas!

**Comment by boddhisattva** [comment]
@kytrinyx Just thinking out loud, in the current Exercism platform, if we have a feature by which we can filter those solutions that have more conversations(in a descending order), would that help? IMHO, it may be one way to address this but there could be things I may have not considered. It would be really nice to get your thoughts on this. Thank you.

**Comment by kytrinyx** [comment]
@boddhisattva This would find conversations where there are lots of comments, but wouldn't help find conversations with interesting content.

One of the risks of surfacing conversations with lots of comments is that these would gain even more comments, which would feed into these same exercises getting more and more attention (often at the expense of other, equally deserving, exercises).

**Comment by boddhisattva** [comment]
@kytrinyx I see where you are coming from but this may be just one side of the coin? For the `sublist` exercise that I solved in Elixir track, the number of comments in @stoft's solution [here](http://exercism.io/submissions/8568240e200c7767cb69bd4f), and in the solution of some other people like Henrik's solution [here](http://exercism.io/submissions/3da33a04239cfd275f5142c0) and Lennart's solution [here](http://exercism.io/submissions/2f4929f3ebc9e26c0f073a3e) made me more curious to look further to see what I could learn from their conversations around code and what tradeoffs they considered.

Also, I'm thinking to find interesting solutions, we could probably consider doing some kind of search filter algorithm based on words in the comments like `readability`, `abstraction`, `struggling`, `good point`, `nice solution`, `suggest(ion)` etc., May be that could help ?

**Comment by stoft** [comment]
@boddhisattva , how did you find our conversations considering they're more than 3 years old?

@kytrinyx I started thinking about various metrics such as comment upvotes, comment length, number of iterations, keywords, randomness, time, and combinations thereof. Even the possibility of hiding long conversations until you yourself comment. The drawback to basically any metric one can think of is that it will basically self-reinforce itself so to speak.

Some thoughts:

* Comment upvotes will skew towards those who get votes. A cap on upvotes may help. Upvotes with expiry could possibly also help but feels iffy.
* Comment length will skew towards length. Not counting length after X chars may help.
* Keywords will skew towards comments with those words, and is not language agnostic. A word cloud may help.
* Randomness on top of one or several of the above may help surface comments that haven't been seen before.
* A combined metric of upvotes/length/keywords and age plus a certain randomness may help in allowing new comments to gain attention.

Considering all we know right now is hypothetical, would it be possible to run a time-boxed experiment to learn more? E.g. implement something of the above, run it live for say two weeks and then evaluate.

**Comment by kytrinyx** [comment]
> would it be possible to run a time-boxed experiment to learn more

Yes, absolutely.

The bottom line is: a lot of things are changing in v2, we are going to need to let things settle and learn from what we've got before experimenting with anything else, but there are a number of things we've been thinking about and evaluating to learn more about this.

**Comment by stoft** [comment]
Another thought that struck me was the option of alternating between various behaviors, highlighting different aspects, "blast from the past", "needs comment", "liked by many" etc.

**Comment by boddhisattva** [comment]
@stoft Sorry for not being able to get back to you earlier on this. I had been following Henrik on Twitter from the Ruby community since a few years back and I started with seeing his Ruby solutions on Exercism. Basically, I had identified a few people who were regularly submitting solutions(in Ruby) and having conversations on Exercism. When I started learning Elixir, I observed a number of people from Stockholm were already on to Elixir for a relatively longer period and since I knew Henrik was from there as well, I started looking up his solutions and I think from thereon I was looking into your and Lennart's  conversations wrt his solutions and then eventually looking into all of your solutions. One thing that has worked for me so far is identifying some people who are contributing to each community regularly via open source or blogs etc., and looking up their solutions is what I usually tend to do first.

**Comment by stoft** [comment]
@boddhisattva excellent feedback! This shows how you acted and the fact that you were "following" a specific user. 👍

**Comment by tleen** [comment]
@boddhisattva I do that too. I find that if I get into a groove doing a track (an exercise a day or a few a week) I tend to move along the track with a group of people. LIke we are all fish in a [school](https://en.wikipedia.org/wiki/Shoaling_and_schooling) moving along the stream at the same time. I tend to look at the results of my school-mates first.

**Comment by boddhisattva** [comment]
Thank you @stoft 🙏 and I'm glad to know you do that too @tleen :). This approach has really helped me when I had to do things at my own pace and it also helped me to go deep. I had bookmarked the exercism profiles of a few users whose solutions I'd regularly see after completing an exercise wrt a particular track. This is how I used to get my mentorship, even though if it wasn't in an active form.
